### ISSUE TITLE
Question - Internals - Remove Parameter O from Algebra. 

### ISSUE THREAD

**Issue by diesalbla** [functional-streams-for-scala/fs2#1361]
Currently, the `fs2.Stream[F[_], O]` data type is implemented as a `FreeC[Algebra[F, O, ?], Unit]`, where
- `FreeC[F[_], R]` is a _free_-monad (+ monad error + interruption?) constructor, 
- `Algebra[F[_], O, R]` is a GADT of basic operations, which are interpreted as a sequence of chained changes to a `CompileScope[F]`, and whose evaluation may output (chunks of) elements of type `O`. 

One interesting aspect of this implementation is that the `O` in itself seems to be orthogonal to the operations of the algebra, which is to say, with the operations involved on evolving the `CompileScope`. 

It would be interesting to find out if that parameter `O`, and the concern it carries (emitting values of the desired type) can be factored out of the `Algebra` and into the `F[_]` type used in there. As an early idea, the behaviour pattern of performing the operations and emitting some output is very similar to the `WriterT` monad transformer.

**Comment by mpilquist** [comment]
The `Output` node seems fairly fundamental to the algebra. Not sure what extracting out `O` buys us hear.

**Comment by diesalbla** [comment]
> The Output node seems fairly fundamental to the algebra. Not sure what extracting out O buys us hear.

If this can be done, what it would buy is to remove the `Output` class from the `Algebra`, and to fold it into the `Eval` class. The action of outputting a `Chunk[O]` would be a _capability_ of the `F[_]` used; capability which with the algebra is not concerned. For the `Algebra`, it would be an `Eval[Unit]`. For the same reason, the `O` in the `Step` would be unknown inside the `Algebra`, and it would only be the code using it, inside the `CompileScope`, which would be using it. 

The intuition here I am trying to follow here is the operation [WriterT.tell](https://github.com/typelevel/cats/blob/master/core/src/main/scala/cats/data/WriterT.scala#L534-L535), in which you indicate that something is sent _on the side_, but which gives no information (`Unit`) to the following operations. In this sense, from the perspective of a `FreeC[ Algebra[F, O, ?], ?]`, the `O` is part of a _writer_ in `F`. 

If this can be done, you would end up with:
- The `Algebra` ADT would be reduced to two parameters: `Algebra[F, R]`. This is what you have now with `AlgEffect`.
- The fact that the `F` is a Writer (of `O` values) would be for the user of the Algebra (the one that fills that `F` parameter) to know.

**Comment by mpilquist** [comment]
That seems like a lot of internal changes and I'm not convinced the benefits are worth it. What practical issue does the extra type parameter cause? How would users benefit from this change? E.g., if we got a big performance boost than this could be worth it.

**Comment by diesalbla** [comment]
Oh, this would have no immediate benefit of those using the `Stream` api. The goal is to (hopefully) make the internal core a bit clearer, by separating independent and orthogonal concerns (resource and scope handling from value emitting values) into separate code parts.

I should mention that I did not open the issue as an immediate concern that needs to be addressed, or may be addressed any time soon. This has no priority at all. I have opened to collect and ask information about what can or cannot be done or not in this regard, and if something cannot be done, then due to which constraints.
